ReactiveCocoa/CocoaTarget.swift

```diff
@@ -4,16 +4,56 @@ import enum Result.NoError
 
 /// A target that accepts action messages.
 internal final class CocoaTarget<Value>: NSObject {
+	private enum State {
+		case idle
+		case sending
+		case replayAfterSending([Value])
+	}
+
 	private let observer: Observer<Value, NoError>
 	private let transform: (Any?) -> Value
-	
+
+	private var state: State
+
 	internal init(_ observer: Observer<Value, NoError>, transform: @escaping (Any?) -> Value) {
 		self.observer = observer
 		self.transform = transform
+		self.state = .idle
 	}
-	
-	@objc internal func sendNext(_ receiver: Any?) {
-		observer.send(value: transform(receiver))
+
+	/// Broadcast the action message to all observers.
+	///
+	/// Reentrancy is supported, and the action message would be deferred until the
+	/// delivery of the current message has completed.
+	///
+	/// - note: It should only be invoked on the main queue.
+	///
+	/// - parameters:
+	///   - sender: The object which sends the action message.
+	@objc internal func invoke(_ sender: Any?) {
+		switch state {
+		case .idle:
+			state = .sending
+			observer.send(value: transform(sender))
+
+			while case let .replayAfterSending(values) = state {
+				if values.count == 1 {
+					state = .sending
+				} else {
+					state = .replayAfterSending(Array(values.dropFirst()))
```

Comment:
This isn't safe.

Individual reads and writes are atomic, but this write depends on the value of the previous read. So they need to be locked together.

Otherwise this read could happen, another thread could append in `replayAfterSending`, then this would remove the first value but unintentionally drop the last value.

It might be worth trying to write a test that catches that as well.